### DISCUSSION TITLE
multiline parsing postfix logs?

### DISCUSSION THREAD

**Discussion by helmut72** [fluent/fluent-bit#7853] [Q&A] [answered]
Hi,

is it possible with multiline parser to parse logs from postfix mail server and glue them together? Common log output from postfix:

```log
Oct 10 15:59:28 mail postfix/smtpd[1830]: connect from example.com[127.0.0.1]
Oct 10 15:59:28 mail postfix/smtpd[1830]: C6E0DDB74006: client=example.com[127.0.0.1]
Oct 10 15:59:28 mail postfix/cleanup[1894]: C6E0DDB74006: message-id=<A40CF64D-7F2D-42E4-8A76-CBFFF64A6EB1@example.com>
Oct 10 15:59:28 mail postfix/qmgr[18719]: C6E0DDB74006: from=<test@example.com>, size=309891, nrcpt=1 (queue active)
Oct 10 15:59:28 mail postfix/smtpd[1830]: disconnect from example.com[127.0.0.1]
Oct 10 15:59:29 mail postfix/smtpd[1827]: connect from example.com[127.0.0.1]
Oct 10 15:59:29 mail postfix/smtpd[1827]: 3D74ADB7400B: client=example.com[127.0.0.1]
Oct 10 15:59:29 mail postfix/cleanup[1695]: 3D74ADB7400B: message-id=<f93388828093534f92d85ffe21b2a719@example.info>
Oct 10 15:59:29 mail postfix/smtpd[1827]: disconnect from example.com[127.0.0.1]
Oct 10 15:59:29 mail postfix/qmgr[18719]: 3D74ADB7400B: from=<test2@example.info>, size=2140, nrcpt=1 (queue active)
Oct 10 15:59:30 mail postfix/smtp[1810]: 3D74ADB7400B: to=<test@example.to>, relay=example.to[192.168.0.20]:25, delay=1.7, delays=0.02/0/1.7/0.06, dsn=2.0.0, status=sent (250 [Sniper] OK 1539154772 snipe-queue 10549)
Oct 10 15:59:30 mail postfix/smtp[1810]: 3D74ADB7400B: to=<test2@example.to>, relay=example.to[192.168.0.20]:25, delay=1.7, delays=0.02/0/1.7/0.06, dsn=2.0.0, status=sent (250 [Sniper] OK 1539154772 snipe-queue 10549)
Oct 10 15:59:30 mail postfix/qmgr[18719]: 3D74ADB7400B: removed
Oct 10 15:59:31 mail postfix/smtpd[1830]: connect from example.com[127.0.0.1]
Oct 10 15:59:31 mail postfix/smtpd[1830]: 6526CDB7400B: client=example.com[127.0.0.1]
Oct 10 15:59:31 mail postfix/cleanup[1894]: 6526CDB7400B: message-id=<153915476795520900002087@example.aaa>
Oct 10 15:59:31 mail postfix/smtpd[1830]: disconnect from example.com[127.0.0.1]
Oct 10 15:59:31 mail postfix/qmgr[18719]: 6526CDB7400B: from=<test@example.bbb>, size=4118, nrcpt=1 (queue active)
Oct 10 15:59:31 mail postfix/smtp[1417]: 6526CDB7400B: to=<test@example.ccc>, relay=example.ccc[192.168.0.10]:25, delay=0.1, delays=0.02/0/0.03/0.05, dsn=2.0.0, status=sent (250 2.0.0 Ok: queued as 71111424269E)
Oct 10 15:59:31 mail postfix/qmgr[18719]: 6526CDB7400B: removed
Oct 10 15:59:32 mail postfix/smtp[1874]: C6E0DDB74006: to=<test@example.ddd>, relay=example.ddd[192.168.0.30]:25, delay=3.4, delays=0.11/0/0.38/2.9, dsn=2.0.0, status=sent (250 2.0.0 OK 1539154772 az9-v6si5976496plb.190 - gsmtp)
Oct 10 15:59:32 mail postfix/qmgr[18719]: C6E0DDB74006: removed
```

All lines with the same ID belongs to the same mail processing. But logs are not ordered. Log lines without an ID are only connection logs. It would be great to get an output like this:

```json
{
  "time": "0000-10-10T15:59:29+09:00",
  "hostname": "mail",
  "process": "postfix/smtpd[1827]",
  "queue_id": "3D74ADB7400B",
  "client_hostname": "example.com",
  "client_ip": "127.0.0.1",
  "message_id": "f93388828093534f92d85ffe21b2a719@example.info",
  "from": "test2@example.info",
  "messages": [
    {
      "time": "0000-10-10T15:59:30+09:00",
      "to": "test@example.to",
      "status": "sent",
      "message": "to=<test@example.to>, relay=example.to[192.168.0.20]:25, delay=1.7, delays=0.02/0/1.7/0.06, dsn=2.0.0, status=sent (250 [Sniper] OK 1539154772 snipe-queue 10549)"
    },
    {
      "time": "0000-10-10T15:59:30+09:00",
      "to": "test2@example.to",
      "status": "sent",
      "message": "to=<test2@example.to>, relay=example.to[192.168.0.20]:25, delay=1.7, delays=0.02/0/1.7/0.06, dsn=2.0.0, status=sent (250 [Sniper] OK 1539154772 snipe-queue 10549)"
    }
  ]
}
```

Source from both:
https://github.com/youyo/postfix-log-parser

It looks like it's possible with logstash:
https://raw.githubusercontent.com/whyscream/postfix-grok-patterns/main/51-filter-postfix-aggregate.conf
But logstash need to much resources and is no option for me.

Is something like this possible with fluent-bit multiline parsing?

Thank you

**Comment by patrick-stephens** [comment, marked as answer]
I don't think normal multiline parsing can do this for you.
You may be able to do it with a more powerful LUA filter or stream processing. 
Or a custom plugin (e.g. Golang) to handle it.

The other thing I was thinking is to parse out the ID then use that in a second stage processing, e.g. send all lines for ID x to a specific file named by the ID, then tail that file and reconstruct as a normal multiline. 

Or use the ID in the tag and get a LUA filter to reconstruct them in the same pipeline maybe?
Or maybe something dirty like running grep via `exec` on the file to just get matching lines for an ID?

**Reply by helmut72** [reply]
Thanks for the suggestions and ideas! I was unsure if this is possible at all.

**Reply by patrick-stephens** [reply]
It's possible, the question is probably more how and where to do it, e.g. a query engine you're sending your logs to (Loki, Elastic, etc.) may be better placed to do it too as it can scale resources then independently as required.

**Reply by helmut72** [reply]
I send it to Loki with json, thanks to fluentbit. Also filtering is possible in Grafana, thanks to the same ID. But this is limited to every single mail.
It's impossible to show senders mail address **and** all his source ip addresses, because both informations are in different log lines (line 1 and 4 in the example). The really interesting filtering (or alerts) are impossible in this case. For example if sender ip addresses are 99% from a specific country, then an alert will be fired if sender ip address comes from a different country. Therefore I want relevant informations in one log line.

Also I want to have some real advantages from this tools and not only some (slow) fancy web-ui. For me, investigate specific problems is always faster in the shell, using journalctl with some grep/awk. With Grafana I want to see the "big picture" of the infrastructure.

> a query engine you're sending your logs to (Loki, Elastic, etc.) may be better placed to do it too as it can scale resources

If the VM where Loki (and Grafana) runs, should also handle all log processing from 100s of other VMs, then they may need to scale. But if proccessing is done on every single VM **before** the logs and metrics will be sent to Loki, I don't think much scaling is needed.
